opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java

```diff
@@ -58,38 +79,79 @@
+            Supplier<ContextWithKey> encryptedContext
     ) {
         if (null != authSession) {
             return authSession.getParent();
         } else {
-            return findServiceSessionByIdOrCreate(request);
+            return findServiceSessionByIdOrCreate(request, encryptedContext);
         }
     }
 
     @NotNull
     @SneakyThrows
-    private <T extends FacadeServiceableGetter> ServiceSession findServiceSessionByIdOrCreate(T request) {
-        if (null != request.getFacadeServiceable().getServiceSessionId()) {
-            Optional<ServiceSession> existingSession = serviceSessions
-                    .findById(request.getFacadeServiceable().getServiceSessionId());
-            if (existingSession.isPresent()) {
-                return existingSession.get();
+    private <T extends FacadeServiceableGetter> ServiceSession findServiceSessionByIdOrCreate(
+            T request,
+            Supplier<ContextWithKey> contextWithKey
+    ) {
+        FacadeServiceableRequest facadeServiceable = request.getFacadeServiceable();
+        UUID serviceSessionId = facadeServiceable.getServiceSessionId();
+
+        ServiceSession session = new ServiceSession();
+        if (null != serviceSessionId) {
+            session.setId(serviceSessionId);
+        }
+
+        session.setContext(contextWithKey.get().getEncryptedContext());
+        session.setFintechOkUri(facadeServiceable.getFintechRedirectUrlOk());
+        session.setFintechNokUri(facadeServiceable.getFintechRedirectUrlNok());
+        session.setSecretKey(secretKeyService.encrypt(contextWithKey.get().getKeyDTO().getKey()));
+        session.setAlgo(properties.getAlgorithm());
+        session.setSalt(contextWithKey.get().getKeyDTO().getSalt());
+        session.setIterCount(properties.getIterationCount());
+        return serviceSessions.save(session);
+    }
+
+    private KeyDTO getSessionSecretKey(UUID serviceSessionId, String sessionPassword) {
+
+        // new secret key
+        if (null == serviceSessionId) {
+            if (Strings.isNullOrEmpty(sessionPassword)) {
+                throw new RuntimeException("No password. Can't generate secret key");
             }
+            byte[] salt = getNewSalt(properties.getSaltLength());
+            byte[] key = secretKeyService.generateKey(sessionPassword, salt);
+            return new KeyDTO(key, salt);
         }
 
-        ServiceSession session = new ServiceSession();
+        Optional<ServiceSession> existingSession = serviceSessions.findById(serviceSessionId);
+        if (!existingSession.isPresent()) {
+            throw new RuntimeException("Session not found");
+        }
 
-        if (null != request.getFacadeServiceable().getServiceSessionId()) {
-            session.setId(request.getFacadeServiceable().getServiceSessionId());
+        // existing saved key
+        byte[] secretKey = existingSession.get().getSecretKey();
+        if (!ArrUtils.isEmpty(secretKey)) {
```

Comment:
secretKey can't be null, it is `@Column(nullable = false)`

Reply:
How then we can clear secretKey in db after some time?

Reply:
that means '@Column(nullable = false)' should be removed

Reply:
Use some clear indication of column state - either column is empty then it is NULL or column is empty - then it is empty byte array